lib/doc/pivot-table.js

```diff
@@ -0,0 +1,132 @@
+  // ExcelJS is >=8.3.0 (as of 2023-10-08).
+  const nameToIndex = cacheFields.reduce((result, cacheField, index) => {
+    result[cacheField.name] = index;
+    return result;
+  }, {});
+  rows = rows.map(row => nameToIndex[row]);
+  columns = columns.map(column => nameToIndex[column]);
+  values = values.map(value => nameToIndex[value]);
+
+  // form pivot table object
+  return {
+    sourceSheet,
+    rows,
+    columns,
+    values,
+    metric: 'sum',
+    cacheFields,
+    // defined in <pivotTableDefinition> of xl/pivotTables/pivotTable1.xml;
+    // also used in xl/workbook.xml
+    cacheId: '10',
+  };
+}
+
+function validate(worksheet, model) {
+  if (worksheet.workbook.pivotTables.length === 1) {
+    throw new Error(
+      'A pivot table was already added. At this time, ExcelJS supports at most one pivot table per file.'
+    );
+  }
+
+  if (model.metric && model.metric !== 'sum') {
+    throw new Error('Only the "sum" metric is supported at this time.');
+  }
+
+  const headerNames = model.sourceSheet.getRow(1).values.slice(1);
+  const isInHeaderNames = objectFromProps(headerNames, true);
+  for (const name of [...model.rows, ...model.columns, ...model.values]) {
+    if (!isInHeaderNames[name]) {
+      throw new Error(`The header name "${name}" was not found in ${model.sourceSheet.name}.`);
+    }
+  }
+
+  if (!model.rows.length) {
+    throw new Error('No pivot table rows specified.');
+  }
+
+  if (!model.columns.length) {
+    throw new Error('No pivot table columns specified.');
+  }
+
+  if (model.values.length !== 1) {
+    throw new Error('Exactly 1 value needs to be specified at this time.');
+  }
+}
+
+function makeCacheFields(worksheet, fieldNamesWithSharedItems) {
+  // Cache fields are used in pivot tables to reference source data.
+  //
+  // Example
+  // -------
+  // Turn
+  //
+  //  `worksheet` sheet values [
+  //    ['A', 'B', 'C', 'D', 'E'],
+  //    ['a1', 'b1', 'c1', 4, 5],
+  //    ['a1', 'b2', 'c1', 4, 5],
+  //    ['a2', 'b1', 'c2', 14, 24],
+  //    ['a2', 'b2', 'c2', 24, 35],
+  //    ['a3', 'b1', 'c3', 34, 45],
+  //    ['a3', 'b2', 'c3', 44, 45]
+  //  ];
+  //  fieldNamesWithSharedItems = ['A', 'B', 'C'];
+  //
+  // into
+  //
+  //  [
+  //    { name: 'A', sharedItems: ['a1', 'a2', 'a3'] },
+  //    { name: 'B', sharedItems: ['b1', 'b2'] },
+  //    { name: 'C', sharedItems: ['c1', 'c2', 'c3'] },
+  //    { name: 'D', sharedItems: null },
+  //    { name: 'E', sharedItems: null }
+  //  ]
+
+  const names = worksheet.getRow(1).values;
+  const nameToHasSharedItems = objectFromProps(fieldNamesWithSharedItems, true);
+
+  const aggregate = columnIndex => {
+    const columnValues = worksheet.getColumn(columnIndex).values.splice(2);
+    const columnValuesAsSet = new Set(columnValues);
+    return toSortedArray(columnValuesAsSet);
+  };
+
+  // make result
+  const result = [];
+  for (const columnIndex of range(1, names.length)) {
```

Comment:
wdyt?
```suggestion
  for (const columnIndex of names.keys()) {
```

Reply:
This won't work, because the iterator will include the index 0 and we don't want that. (This is because of the design of `.getRow()` where we start counting at index 1 vs. index 0 which is customary otherwise in JavaScript.)

```js
> Array.from(worksheet1.getRow(1).values.keys())
[
  0, 1, 2, 3, 4,
  5, 6, 7, 8
]
```